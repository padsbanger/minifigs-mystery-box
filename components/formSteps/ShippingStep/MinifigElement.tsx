import Heading from "@/components/ui/Heading/Heading";
import { MinifigPart } from "@/lib/utils/services/MinifigsService/MinifigsService.types";
import Image from "next/image";

const MinifigElement = ({ part }: MinifigPart) => {
  return (
    <div className="flex flex-col/ mb-4 justify-start gap-2">
      <Image
        src={part.part_img_url}
        alt={part.name}
        width="64"
        height="64"
        placeholder="blur"
        blurDataURL="/placeholder.jpg"
      />
      <div className="flex flex-col">
        <Heading
          variant="heading8"
          level={5}
          text={part.name}
          classes="font-bold line-clamp-1"
        />
        <Heading
          variant="heading9"
          level={6}
          text={part.part_num}
          classes="font-bold line-clamp-1 text-blue-500"
        />
      </div>
    </div>
  );
};

export default MinifigElement;
