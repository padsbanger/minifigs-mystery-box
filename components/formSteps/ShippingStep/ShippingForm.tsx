import Heading from "@/components/ui/Heading/Heading";

const ShippingForm = () => {
  return (
    <div className="basis-2/3">
      <Heading
        text="Shipping Details"
        variant={"heading3"}
        level={3}
        classes="my-4"
      />
    </div>
  );
};

export default ShippingForm;
