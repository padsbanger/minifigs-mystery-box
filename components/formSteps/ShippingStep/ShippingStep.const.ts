import * as yup from "yup";

const PHONE_REGEXP =
  /^((\\+[1-9]{1,4}[ \\-]*)|(\\([0-9]{2,3}\\)[ \\-]*)|([0-9]{2,4})[ \\-]*)*?[0-9]{3,4}?[ \\-]*[0-9]{3,4}?$/;

export const SHIPPING_FORM_SCHEMA = yup
  .object({
    firstName: yup.string().required("Please provide name."),
    surName: yup.string().required("Please provide surname."),
    phoneNumber: yup
      .string()
      .matches(PHONE_REGEXP, "Phone number is not valid.")
      .required("Please provide phone number."),
    email: yup.string().email().required("Please provide valid email."),
    dateOfBirth: yup
      .date()
      .nullable()
      .transform((curr, orig) => (orig === "" ? null : curr))
      .required("Please provide date of birth."),
    address: yup.string().required("Please provide address."),
    city: yup.string().required("Please provide city."),
    state: yup.string().required("Please provide state."),
    zip: yup
      .string()
      .required("Please provide zip code.")
      .matches(/^[0-9]{5}$/, "Must be exactly 5 digits."),
  })
  .required();
