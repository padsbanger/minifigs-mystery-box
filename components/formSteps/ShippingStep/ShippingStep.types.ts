export type ShippingFormValues = {
  firstName: string;
  surName: string;
  phoneNumber: string;
  email: string;
  dateOfBirth: Date;
  address: string;
  city: string;
  state: string;
  zip: string;
};
