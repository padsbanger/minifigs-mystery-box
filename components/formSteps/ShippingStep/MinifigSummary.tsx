import Heading from "@/components/ui/Heading/Heading";
import Image from "next/image";
import { useEffect } from "react";

import {
  useIsLoadingMinifigs,
  useMinifigsStore,
  useSelectedMinifig,
  useSelectedMinifigParts,
} from "@/lib/utils/services/MinifigsService/useMinifigsStore/useMinifigsStore";
import MinifigElement from "./MinifigElement";
import type { MinifigPart } from "@/lib/utils/services/MinifigsService/MinifigsService.types";
import Spinner from "@/components/ui/Spinner/Spinner";
import Button from "@/components/ui/Button/Button";
import clsx from "clsx";

type MinifigSummaryProps = {
  handleSubmit: () => void;
  isSubmitting: boolean;
};

const MinifigSummary = ({
  handleSubmit,
  isSubmitting,
}: MinifigSummaryProps) => {
  const getMinifigParts = useMinifigsStore.getState().getMinifigParts;
  const isLoading = useIsLoadingMinifigs();
  const selectedMinifig = useSelectedMinifig();
  const minifigParts = useSelectedMinifigParts();

  useEffect(() => {
    if (selectedMinifig?.set_num) {
      getMinifigParts(selectedMinifig?.set_num);
    }
  }, []);

  return (
    <div
      className={clsx(
        "flex flex-col bg-white shadow-md rounded-md basis-1/3 p-4 h-[calc(100vh-20rem)] min-w-[320px] overflow-auto",
        { "opacity-70": isSubmitting || isLoading },
      )}
    >
      <Heading
        text="Summary"
        variant={"heading3"}
        level={3}
        classes="my-4 text-black font-bold"
      />
      <Image
        src={selectedMinifig?.set_img_url as string}
        alt={selectedMinifig?.name as string}
        width="180"
        height="120"
        className="m-auto object-cover my-4 w-[180px]"
      />
      <Heading
        text={selectedMinifig?.name as string}
        variant={"heading8"}
        level={6}
        classes="my-4 text-black text-center font-bold"
      />
      {isLoading ? (
        <Spinner />
      ) : (
        <>
          <p className="font-plus-jakarta-sans my-4">
            There are {minifigParts.length} parts in this minifig:
          </p>
          <div className="h-auto md:h-[calc(100vh-1rem)] overflow-auto">
            {minifigParts.map((el: MinifigPart) => (
              <MinifigElement key={el.id} {...el} />
            ))}
          </div>
        </>
      )}
      <div>
        <Button
          type="submit"
          label="Submit"
          classes="flex items-center h-10 mx-auto my-10"
          onClick={handleSubmit}
          isLoading={isSubmitting}
          disabled={isSubmitting || isLoading}
        />
      </div>
    </div>
  );
};

export default MinifigSummary;
