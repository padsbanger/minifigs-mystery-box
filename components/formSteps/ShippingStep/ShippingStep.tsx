import Heading from "@/components/ui/Heading/Heading";
import MinifigSummary from "./MinifigSummary";
import { yupResolver } from "@hookform/resolvers/yup";
import { SubmitHandler, useForm } from "react-hook-form";
import { SHIPPING_FORM_SCHEMA } from "./ShippingStep.const";
import { ShippingFormValues } from "./ShippingStep.types";
import TextInput from "@/components/ui/TextInput/TextInput";
import {
  useMinifigsStore,
  useIsSubmittingOrder,
} from "@/lib/utils/services/MinifigsService/useMinifigsStore/useMinifigsStore";

const ShippingStep = () => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<ShippingFormValues>({
    reValidateMode: "onChange",
    resolver: yupResolver(SHIPPING_FORM_SCHEMA),
  });
  const saveOrder = useMinifigsStore.getState().saveOrder;
  const isSubmittingOrder = useIsSubmittingOrder();
  const onSubmit: SubmitHandler<ShippingFormValues> = (data) => {
    saveOrder(data);
  };

  return (
    <div className="flex flex-col items-center md:h-screen h-auto justify-center">
      <div className="container mx-auto">
        <form onSubmit={handleSubmit(onSubmit)}>
          <div className="flex flex-col md:flex-row gap-10">
            <div className="basis-2/3">
              <Heading
                text="Shipping Details"
                variant={"heading3"}
                level={3}
                classes="my-4 text-white"
              />
              <div className="grid grid-cols-2 gap-4">
                <TextInput
                  label="First Name"
                  placeholder="Please provide your name"
                  {...register("firstName")}
                  errorMessage={errors.firstName?.message}
                  disabled={isSubmittingOrder}
                />
                <TextInput
                  label="Surname"
                  placeholder="Please provide your last name"
                  {...register("surName")}
                  errorMessage={errors.surName?.message}
                  disabled={isSubmittingOrder}
                />
              </div>
              <TextInput
                label="Phone number"
                placeholder="Please provide your phone number"
                {...register("phoneNumber")}
                errorMessage={errors.phoneNumber?.message}
                disabled={isSubmittingOrder}
              />
              <TextInput
                label="Email"
                placeholder="Please provide your email"
                {...register("email")}
                errorMessage={errors.email?.message}
                disabled={isSubmittingOrder}
              />
              <TextInput
                label="Date of birth"
                type="date"
                placeholder="Please provide your date of birth"
                {...register("dateOfBirth")}
                errorMessage={errors.dateOfBirth?.message}
                disabled={isSubmittingOrder}
              />
              <TextInput
                label="Address"
                {...register("address")}
                placeholder="Please provide your address"
                errorMessage={errors.address?.message}
                disabled={isSubmittingOrder}
              />
              <TextInput
                label="City"
                placeholder="Please provide your city"
                {...register("city")}
                errorMessage={errors.city?.message}
                disabled={isSubmittingOrder}
              />
              <div className="grid grid-cols-2 gap-4">
                <TextInput
                  label="State"
                  placeholder="Please provide your state"
                  {...register("state")}
                  errorMessage={errors.state?.message}
                  disabled={isSubmittingOrder}
                />
                <TextInput
                  placeholder="Please provide your zip code"
                  label="Zip Code"
                  type="number"
                  {...register("zip")}
                  errorMessage={errors.zip?.message}
                  disabled={isSubmittingOrder}
                />
              </div>
            </div>
            <MinifigSummary
              handleSubmit={() => handleSubmit(onSubmit)}
              isSubmitting={isSubmittingOrder}
            />
          </div>
        </form>
      </div>
    </div>
  );
};

export default ShippingStep;
