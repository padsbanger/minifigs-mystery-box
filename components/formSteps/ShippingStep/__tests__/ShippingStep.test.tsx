import { render } from "@testing-library/react";
import ShippingStep from "../ShippingStep";

const setup = () => render(<ShippingStep />);

describe("ShippingStep View", () => {
  test("should match snapshot", () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });
});
