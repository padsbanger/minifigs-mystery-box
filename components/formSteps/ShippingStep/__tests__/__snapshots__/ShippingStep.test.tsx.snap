// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShippingStep View should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-center md:h-screen h-auto justify-center"
  >
    <div
      class="container mx-auto"
    >
      <form>
        <div
          class="flex flex-col md:flex-row gap-10"
        >
          <div
            class="basis-2/3"
          >
            <h3
              class="font-plus-jakarta-sans leading-normal uppercase text-headline-3 font-700 leading-[130%] my-4 text-white"
            >
              Shipping Details
            </h3>
            <div
              class="grid grid-cols-2 gap-4"
            >
              <div
                class="flex flex-col my-1"
              >
                <label
                  class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                  for="firstName"
                >
                  First Name
                </label>
                <input
                  class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                  id="firstName"
                  name="firstName"
                  placeholder="Please provide your name"
                />
              </div>
              <div
                class="flex flex-col my-1"
              >
                <label
                  class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                  for="surName"
                >
                  Surname
                </label>
                <input
                  class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                  id="surName"
                  name="surName"
                  placeholder="Please provide your last name"
                />
              </div>
            </div>
            <div
              class="flex flex-col my-1"
            >
              <label
                class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                for="phoneNumber"
              >
                Phone number
              </label>
              <input
                class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                id="phoneNumber"
                name="phoneNumber"
                placeholder="Please provide your phone number"
              />
            </div>
            <div
              class="flex flex-col my-1"
            >
              <label
                class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                for="email"
              >
                Email
              </label>
              <input
                class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                id="email"
                name="email"
                placeholder="Please provide your email"
              />
            </div>
            <div
              class="flex flex-col my-1"
            >
              <label
                class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                for="dateOfBirth"
              >
                Date of birth
              </label>
              <input
                class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                id="dateOfBirth"
                name="dateOfBirth"
                placeholder="Please provide your date of birth"
                type="date"
              />
            </div>
            <div
              class="flex flex-col my-1"
            >
              <label
                class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                for="address"
              >
                Address
              </label>
              <input
                class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                id="address"
                name="address"
                placeholder="Please provide your address"
              />
            </div>
            <div
              class="flex flex-col my-1"
            >
              <label
                class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                for="city"
              >
                City
              </label>
              <input
                class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                id="city"
                name="city"
                placeholder="Please provide your city"
              />
            </div>
            <div
              class="grid grid-cols-2 gap-4"
            >
              <div
                class="flex flex-col my-1"
              >
                <label
                  class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                  for="state"
                >
                  State
                </label>
                <input
                  class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                  id="state"
                  name="state"
                  placeholder="Please provide your state"
                />
              </div>
              <div
                class="flex flex-col my-1"
              >
                <label
                  class="font-plus-jakarta-sans cursor-pointer text-white my-1 p-1"
                  for="zip"
                >
                  Zip Code
                </label>
                <input
                  class="font-plus-jakarta-sans p-2 active:shadow-lg focus:shadow-lg rounded-lg outline-none transition-shadow"
                  id="zip"
                  name="zip"
                  placeholder="Please provide your zip code"
                  type="number"
                />
              </div>
            </div>
          </div>
          <div
            class="flex flex-col bg-white shadow-md rounded-md basis-1/3 p-4 h-[calc(100vh-20rem)] min-w-[320px] overflow-auto"
          >
            <h3
              class="font-plus-jakarta-sans leading-normal uppercase text-headline-3 font-700 leading-[130%] my-4 text-black font-bold"
            >
              Summary
            </h3>
            <img
              class="m-auto object-cover my-4 w-[180px]"
              data-nimg="1"
              decoding="async"
              height="120"
              src=""
              style="color: transparent;"
              width="180"
            />
            <h6
              class="font-plus-jakarta-sans leading-normal uppercase text-headline-8 font-700 leading-[145%] my-4 text-black text-center font-bold"
            />
            <p
              class="font-plus-jakarta-sans my-4"
            >
              There are 0 parts in this minifig:
            </p>
            <div
              class="h-auto md:h-[calc(100vh-1rem)] overflow-auto"
            />
            <div>
              <button
                class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-6 px-12 rounded-md disabled:opacity-55 shadow-sm hover:shadow-xl transition-all active:shadow-2xl flex items-center h-10 mx-auto my-10"
              >
                <span
                  class="uppercase font-plus-jakarta-sans"
                >
                  Submit
                </span>
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
